File: typescript/src/utils/batch.ts
import { MpesaApiClient } from "../client/client.js";

export interface BatchRequest {
  method: "POST" | "GET";
  url: string;
  data?: unknown;
}

export async function executeBatch(
  client: MpesaApiClient,
  requests: BatchRequest[],
  concurrency = 3,
): Promise<unknown[]> {
  const results: unknown[] = [];
  for (let i = 0; i < requests.length; i += concurrency) {
    const chunk = requests.slice(i, i + concurrency);
    const chunkResults = await Promise.all(
      chunk.map((req) => {
        if (req.method === "POST") {
          return client.post(req.url, req.data);
        }
        return client.get(req.url);
      }),
    );
    results.push(...chunkResults);
  }
  return results;
}

File: go/client/batch.go
package client

import (
	"context"
)

type BatchRequest struct {
	Method string
	URL    string
	Body   interface{}
}

type BatchResult struct {
	Data []byte
	Err  error
}

func (c *Client) ExecuteBatch(ctx context.Context, requests []BatchRequest, concurrency int) []BatchResult {
	if concurrency <= 0 {
		concurrency = 3
	}

	results := make([]BatchResult, 0, len(requests))

	for i := 0; i < len(requests); i += concurrency {
		end := i + concurrency
		if end > len(requests) {
			end = len(requests)
		}
		chunk := requests[i:end]

		type chResult struct {
			index int
			data  []byte
			err   error
		}
		ch := make(chan chResult, len(chunk))

		for _, req := range chunk {
			r := req
			go func() {
				data, err := c.doRequest(ctx, r.Method, r.URL, r.Body)
				ch <- chResult{data: data, err: err}
			}()
		}

		for range chunk {
			res := <-ch
			results = append(results, BatchResult{Data: res.data, Err: res.err})
		}
	}

	return results
}
